### PR TITLE
subsystem: fpga_bridge: Added FPGA Bridge

### PR DESCRIPTION
**Pull request by hardeepsharma95**
Added FPGA Bridge Enable/Disable functionality :

- Adding support for Agilex & Agilex5
- Adding Ztest for FPGA bridge enable/disable

### CODE REVIEW

File: samples/subsys/shell/shell_module/boards/intel_socfpga_agilex5_socdk.conf
Comment (kartben) [inline samples/subsys/shell/shell_module/boards/intel_socfpga_agilex5_socdk.conf:1]: The purpose of the shell module sample is not to demo other shell modules but rather to show how to write one
You should consider creating a dedicated sample .

File: CODEOWNERS
Comment (kartben) [inline CODEOWNERS:1]: Shouldn't be in the commit

**Review by kartben — CHANGES_REQUESTED** [review]

**Comment by kartben** [comment]
Sorry I inadvertently marked as ready for review by misclicking on my phone.

File: samples/subsys/shell/shell_module/boards/intel_socfpga_agilex5_socdk.conf
Comment (hardeepsharma95) [reply samples/subsys/shell/shell_module/boards/intel_socfpga_agilex5_socdk.conf:1]: Could you please eloborate more I didn't understand what do you mean by demo shell module here ?
The bridge shell is a generic shell not intel specific.

Comment (kartben) [reply samples/subsys/shell/shell_module/boards/intel_socfpga_agilex5_socdk.conf:1]: Hi @hardeepsharma95  -- the `shell_module` sample is here for demonstrating how to use the Zephyr shell API to add new commands etc.
Please consider adding a new sample in `samples/subsys/fpga_bridge/shell` similar to e.g. `samples/subsys/usb/shell` and have it "just" enable a shell and use the README to detail some of the things folks can do using the available commands.
Does this help?

Comment (hardeepsharma95) [reply samples/subsys/shell/shell_module/boards/intel_socfpga_agilex5_socdk.conf:1]: FPGA bridge shell implementation is here subsys/fpga_bridge/fpga_bridge_shell.c
We are just adding support for intel_socfpga_agilex5_socdk board we are not changing the prj.conf.
Other peripheral like counter (CONFIG_COUNTER_SHELL) and sip_svc (CONFIG_ARM_SIP_SVC_SUBSYS_SHELL) also following the same approach

File: CODEOWNERS
Comment (hardeepsharma95) [reply CODEOWNERS:1]: Hi @carlescufi @nashif @jakub-uC Can you please review this PR ?

**Review by hardeepsharma95 — COMMENTED** [review]
Hi [@nashif](https://github.com/nashif) [@marcinszkudlinski](https://github.com/marcinszkudlinski) [@softwarecki](https://github.com/softwarecki) [@carlocaione](https://github.com/carlocaione) [@povergoing](https://github.com/povergoing) Can you please review this PR ?

File: samples/subsys/shell/shell_module/boards/intel_socfpga_agilex5_socdk.conf
Comment (kartben) [reply samples/subsys/shell/shell_module/boards/intel_socfpga_agilex5_socdk.conf:1]: > FPGA bridge shell implementation is here subsys/fpga_bridge/fpga_bridge_shell.c 

Yes, I understand. But this sample is meant to demo how to use the Shell API to add new commands, and not for showcasing specific shell modules. Please look at `samples/subsys/usb/shell` or `samples/subsys/display/cfb_shell` as examples of how you could add a sample specifically focused on helping folks exercise the FPGA bridge shell commands.

> We are just adding support for intel_socfpga_agilex5_socdk board we are not changing the prj.conf. 
> Other peripheral like counter (CONFIG_COUNTER_SHELL) and sip_svc (CONFIG_ARM_SIP_SVC_SUBSYS_SHELL) also following the same approach

That was IMO introduced by mistake.
If you need some kind of toolbox containing all the "shell troubleshooting commands" for intel_socfpga_agilex5_socdk you may want to consider creating a [snippet](https://docs.zephyrproject.org/latest/build/snippets/index.html) with the relevant conf and overlay bits.

Comment (kartben) [reply samples/subsys/shell/shell_module/boards/intel_socfpga_agilex5_socdk.conf:1]: @hardeepsharma95 just making sure you saw this follow-up comment. Thanks!

**Comment by kartben** [comment]
btw was this new subsystem presented to / discussed in the architecture working group?

**Comment by hardeepsharma95** [comment]
> btw was this new subsystem presented to / discussed in the architecture working group?

No I have not disscussed but if required I can present in the architecture working group

File: samples/subsys/shell/shell_module/boards/intel_socfpga_agilex5_socdk.conf
Comment (tanmaykathpalia) [reply samples/subsys/shell/shell_module/boards/intel_socfpga_agilex5_socdk.conf:1]: @kartben I have created a [snippet ](https://github.com/zephyrproject-rtos/zephyr/pull/67660) PR for the same as you mentioned and removed the .conf file from the shell module sample.
Thanks for pointing toward the snippet and would be helpful if you could review it and provide your valuable feedback.

Comment (hardeepsharma95) [reply samples/subsys/shell/shell_module/boards/intel_socfpga_agilex5_socdk.conf:1]: > @hardeepsharma95 just making sure you saw this follow-up comment. Thanks!

Thanks @kartben :-) for the suggestion I have removed the changes from the shell_module and we have created the snippets PR for our specific board.

**Review by hardeepsharma95 — COMMENTED** [review]
Hi @kartben [@nashif](https://github.com/nashif) [@marcinszkudlinski](https://github.com/marcinszkudlinski) [@softwarecki](https://github.com/softwarecki) [@carlocaione](https://github.com/carlocaione) [@povergoing](https://github.com/povergoing) Can you please review this PR ?

**Comment by kartben** [comment]
> > btw was this new subsystem presented to / discussed in the architecture working group?
> 
> No I have not disscussed but if required I can present in the architecture working group

Yes, I think it would be appropriate. @nashif or @carlescufi can probably confirm.
I am not really familiar with the domain but I have to admit that introducing a whole new subsystem for just one API seems odd but I'm probably missing something :)

Thanks!

**Review by hardeepsharma95 — COMMENTED** [review]
Hi @kartben [@nashif](https://github.com/nashif) [@marcinszkudlinski](https://github.com/marcinszkudlinski) [@softwarecki](https://github.com/softwarecki) [@carlocaione](https://github.com/carlocaione) [@povergoing](https://github.com/povergoing) Can you please review this PR ?

**Review by galak — COMMENTED** [review]
1. Is this a concept that exists in other FPGA from AMD/Xilinx, Lattice, etc?
2. Does it make sense to use fpga_bridge_intel.{c.h} vs fpga_bridge_agilex?  (only ask with the intent for Intel to spin out the FPGA business, so guessing things may get renamed)

**Comment by henrikbrixandersen** [comment]
> 1. Is this a concept that exists in other FPGA from AMD/Xilinx, Lattice, etc?

Yes, at least the Xilinx Zynq 7000 series have a similar concept.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Shouldn't turning on/off the "bridges", "level converters" or whatever term is used by a given vendor for the logic connecting the CPU part and the FPGA part of the SoC just be handled as part of the existing API in https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/fpga.h for loading a bitstream?

This is how the Linux kernel does this, see e.g. https://elixir.bootlin.com/linux/latest/source/drivers/fpga/zynq-fpga.c#L289

**Comment by hardeepsharma95** [comment]
> Shouldn't turning on/off the "bridges", "level converters" or whatever term is used by a given vendor for the logic connecting the CPU part and the FPGA part of the SoC just be handled as part of the existing API in https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/fpga.h for loading a bitstream?
> 
> This is how the Linux kernel does this, see e.g. https://elixir.bootlin.com/linux/latest/source/drivers/fpga/zynq-fpga.c#L289

Thanks for your valuable comment but just to clarify in this PR we are not loading any bitstream here we are only  enabling/disabling the any specific bridge or all the bridges

**Comment by hardeepsharma95** [comment]
> Thanks for your valuable commen

Thanks for your valuable comment I have changed the file name from intel to agilex

**Comment by hardeepsharma95** [comment]
> 1. Is this a concept that exists in other FPGA from AMD/Xilinx, Lattice, etc?
> 2. Does it make sense to use fpga_bridge_intel.{c.h} vs fpga_bridge_agilex?  (only ask with the intent for Intel to spin out the FPGA business, so guessing things may get renamed)

Thanks for your valuable comment I have changed the file name from intel to agilex

**Review by hardeepsharma95 — COMMENTED** [review]
Hi @henrikbrixandersen @galak @tanmaykathpalia @kartben @carlescufi @jakub-uC @nashif  can you please review this PR ?

**Comment by henrikbrixandersen** [comment]
> Hi @henrikbrixandersen @galak @tanmaykathpalia @kartben @carlescufi @jakub-uC @nashif can you please review this PR ?

I have already given my review comment in https://github.com/zephyrproject-rtos/zephyr/pull/65753#pullrequestreview-1865454505

I still do not think we should introduce a new API for this. I'd prefer if this could a) be automatically handled by the existing FPGA API or b) handled via new API calls added to the existing FPGA API.

**Comment by tanmaykathpalia** [comment]
> > Hi @henrikbrixandersen @galak @tanmaykathpalia @kartben @carlescufi @jakub-uC @nashif can you please review this PR ?
> 
> I have already given my review comment in [#65753 (review)](https://github.com/zephyrproject-rtos/zephyr/pull/65753#pullrequestreview-1865454505)
> 
> I still do not think we should introduce a new API for this. I'd prefer if this could a) be automatically handled by the existing FPGA API or b) handled via new API calls added to the existing FPGA API.

@kartben  @henrikbrixandersen 
 I am of a different opinion for following reasons, and why we need a separate new API:

1. As there can be multiple ways to connect between FPGA and APU (Application processor unit) based on direction, speed, coherency, etc., like APU to FPGA, FPGA to APU, etc. for Intel we call this connection bridges.  Now, in the current implementation of fpga.h there is no way we can pass the bridge type or do specific bridge enable/disable.

2. Also, for security purpose the direct access to registers is restricted and we need to use the SiP SVC (Silicon Provider) API to read/write to registers using ARM trusted firmware. So, this bridge enable/disable functionality will be a client for SiP SVC mailbox which in turn will be used by different applications to see whether bridges are enabled/disabled, and I think it won't be fit in the driver category as there is no direct communication with the hardware.

For reference in U-Boot code: https://github.com/u-boot/u-boot/blob/master/arch/arm/mach-socfpga/misc.c

File: include/zephyr/fpga_bridge/fpga_bridge.h [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+/**
+ * @brief Do the bridge reset based on action enable/disable
+ *
+ * @param[in] action Enable/Disable bridges
+ * @param[in] mask Bridge mask
+ *
+ * @return 0 on success, negative errno code on fail
+ */
+int32_t do_bridge_reset(uint32_t action, uint32_t mask);
```
Comment (carlescufi) [inline include/zephyr/fpga_bridge/fpga_bridge.h:27]: `fpga_bridge_reset`

**Comment by carlescufi** [comment]
Architecture WG:

- @hardeepsharma95 presents the feature
- @henrikbrixandersen suggests instead extending `include/drivers/fpga.h` instead of creating a new subsystem
- @hardeepsharma95 responds that the FPGA is not a device, it's not a node in the Devicetree. Also, there is a call from non-secure to secure world to reset the bridge
- @nashif agrees that this looks more like a driver than a subsystem. The moment you add hardware-specific implementations it looks more like a driver than a subsystem
- Either a driver or a SoC-specific API

**Comment by hardeepsharma95** [comment]
As discussed in the Architecture group meeting moved this from subsystem to driver.
New PR :: #71360 
Can you please this new PR ?

**Pull request closed without merging**